api/src/wfl/module/covid.clj

```diff
@@ -409,16 +409,40 @@
                      :updated               now})]
       (jdbc/insert-multi! tx details (map to-rows workflows)))))
 
+(defn ^:private update-terra-workflow-statuses!
+  "Update statuses in DETAILS table for active or failed WORKSPACE workflows."
+  [{:keys [workspace details] :as _executor}]
+  (let [query  "SELECT *
+                FROM %s
+                WHERE rawls_submission_id IS NOT NULL
+                AND workflow_id IS NOT NULL
+                AND workflow_status NOT IN ('Succeeded', 'Aborted')"
+        records (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+                  (jdbc/query tx (format query details)))]
+    (letfn [(get-status! [{:keys [rawls_submission_id workflow_id] :as _record}]
+              (:status (firecloud/get-workflow workspace
+                                               rawls_submission_id
+                                               workflow_id)))
+            (update!     [{:keys [id] :as _record} status]
```

Comment:
This approach avoids external calls within a DB transaction but it does have me creating a new transaction for each record updated, not sure if we care.

Reply:
Can we use one for the read and one for the writes?
```clojure
(letfn [(read-active-or-failed-workflows []
          (let [query "your clever query"]
            (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
              (jdbc/query tx [(format query details)]))))
        (get-updated-workflow-status [{:keys [rawls_submission_id workflow_id] :as record}]
          (update
            record
            :workflow_status
            (firecloud/get-workflow workspace rawls_submission_id workflow_id)))
        (write-workflow-statuses [now records]
          (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
            (run!
              (fn [{:keys [id workflow_status]}]
                (jdbc/update! tx details
                  {:workflow_status workflow_status
                   :updated         now}
                  ["id = ?" id]))
              records)))]
  (->> read-active-or-failed-workflows
      (map get-updated-workflow-status)
      (write-workflow-statuses (OffsetDateTime/now))))
```

Reply:
Great, thanks for this suggestion, I thought something like a map might allow this but it helps to see it spelled out. I'll give it a go.

Reply:
Adjusted to follow this suggestion. Thanks!